src/Catalog/CommitCollector.cs

```diff
@@ -56,10 +57,25 @@ public CommitCollector(Uri index, Func<HttpMessageHandler> handlerFunc = null)
                         batch.CommitTimeStamp == lastBatch.CommitTimeStamp, 
                         cancellationToken);
 
-                    front.Value = batch.CommitTimeStamp;
-                    await front.Save(cancellationToken);
+                    // Only commit if this is the last batch that is going to be processed or if the commit timestamp
+                    // has changed from the previous batch. This is important because if two batches have the same
+                    // commit timestamp but processing the second fails, we should not progress the cursor forward.
+                    if (ReferenceEquals(batch, lastBatch))
+                    {
+                        front.Value = batch.CommitTimeStamp;
+                        await front.Save(cancellationToken);
+                        Trace.TraceInformation("CommitCatalog.Fetch front.Value saved due to last batch: {0}", front);
+                    }
+                    else if (previousCommitTimeStamp.HasValue && previousCommitTimeStamp != batch.CommitTimeStamp)
```

Comment:
This won't work either. This will make you always commit after the first batch. Suppose:

ID  | Version  | Commit Timestamp
--  | -------  | ----------------
A   | 1.0.0    | 10:00 PM
B   | 1.0.0    | 10:00 PM
C   | 1.0.0    | 10:05 PM

If you fail on the second batch, you will miss it upon re-entry since the cursor has moved to 10:00 PM.

I think the way the code is written (committing after the batch as complete but having no reference to the next batch) makes it hard to get it right. 

Really, we want to write the cursor _between_ batches. So if your timestamp goes from 10:00 PM to 10:05 PM, you should write the cursor of 10:00 PM before the first 10:05 PM batch is even touched.